### PR TITLE
Switch to FlutterSurfaceView if no Android view is in the frame

### PR DESCRIPTION
**Pull request by blasten**
## Description

This is the last part to address https://github.com/flutter/flutter/issues/58291.  When an Android view is present in the frame, we switch the surface from `FlutterSurfaceView` to `FlutterImageView`.   When no Android view is present, then we switch back to previous surface.

cc @jason-simmons 

## Related Issues

Fixes https://github.com/flutter/flutter/issues/58291

## Tests

I added the following tests: Unit tests (WIP)

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: shell/platform/android/external_view_embedder/external_view_embedder.cc [hunk 1]
```diff
@@ -142,8 +144,14 @@ bool AndroidExternalViewEmbedder::SubmitFrame(
...
   }
   // Submit the background canvas frame before switching the GL context to
-  // the surfaces above.
-  frame->Submit();
+  // the overlay surfaces.
+  //
+  // Skip a frame if the embedding is switching surfaces.
+  auto should_submit_current_frame =
+      previous_frame_view_count_ > 0 || current_frame_view_count == 0;
+  if (should_submit_current_frame) {
```
Comment (blasten) [inline shell/platform/android/external_view_embedder/external_view_embedder.cc:152]: this is `false`, when the Java side is switching from `FlutterSurfaceView` to `FlutterImageView`.

Comment (blasten) [inline shell/platform/android/external_view_embedder/external_view_embedder.cc:152]: This is `false`, when the Java side is switching from `FlutterSurfaceView` to `FlutterImageView`.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java [hunk 2]
```diff
@@ -691,49 +690,98 @@ public void onBeginFrame() {
...
 
   public void onEndFrame() {
-    // Hide overlay surfaces that aren't rendered in the current frame.
+    // Whether the current frame was rendered using ImageReaders.
+    //
+    // Since the image readers may not have images available at this point,
+    // this becomes true if all the required surfaces have images available.
+    //
+    // This is used to decide if the platform views can be rendered in the current frame.
+    // If one of the surfaces don't have an image, the frame may be incomplete and must be dropped.
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:699]: typo: "doesn't"

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java [hunk 3]
```diff
@@ -691,49 +690,98 @@ public void onBeginFrame() {
...
-      if (!currentFrameUsedPlatformViewIds.contains(viewId)) {
-        platformViews.get(viewId).setVisibility(View.GONE);
-        mutatorViews.get(viewId).setVisibility(View.GONE);
+      View platformView = platformViews.get(viewId);
+      View mutatorView = mutatorViews.get(viewId);
+
+      // Show platform views only if the surfaces have images available in this frame,
+      // and if the platform view is rendered in this frame.
+      //
+      // Otherwise, hide the platform view, but don't remove it from view hierarchy yet as
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java:745]: typo: "the view hierarchy"

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java (hunk 2 above)
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: Done.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java (hunk 3 above)
Comment (blasten) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: Done.

**Comment by blasten** [comment]
Linux Web Engine is failing on 
```
/home/swarming/.pub-cache/hosted/pub.dartlang.org/test_core-0.3.4/lib/test_core.dart:69:5: Error: 'unawaited' is imported from both 'package:meta/meta.dart' and 'package:test_api/src/utils.dart'.
    unawaited(Future.error('Dummy exception to set exit code.'));
```

That appears to be fixed upstream.

**Pull request merged**